In [1]:
# Importing required libraries
import numpy as np
import pandas as pd


In [2]:
# Reading the data files into DataFrames
books = pd.read_csv("C:/Users/Navika/Downloads/mlll/Book-Recommendation-System-main/Books.csv")
users = pd.read_csv("C:/Users/Navika/Downloads/mlll/Book-Recommendation-System-main/Users.csv")
ratings = pd.read_csv("C:/Users/Navika/Downloads/mlll/Book-Recommendation-System-main/Ratings.csv")

C:\Users\Navika\AppData\Local\Temp\ipykernel_12808\425881386.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv("C:/Users/Navika/Downloads/mlll/Book-Recommendation-System-main/Books.csv")


In [3]:
# Displaying the first five rows of the books DataFrame
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
# Displaying the first five rows of the users DataFrame
users.head()


User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
# Displaying the first five rows of the ratings DataFrame
ratings.head()


User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [6]:
# Displaying the first five rows of the users DataFrame
users.head()


User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [7]:
# Printing the number of rows and columns in each DataFrame
print(books.shape)
print(users.shape)
print(ratings.shape)


(271360, 8)
(278858, 3)
(1149780, 3)


In [8]:
# Checking for missing values in the books DataFrame
books.isnull().sum()


ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [9]:
# Checking for missing values in the users DataFrame
users.isnull().sum()


User-ID          0
Location         0
Age         110762
dtype: int64

In [10]:
# Checking for missing values in the ratings DataFrame
ratings.isnull().sum()


User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [11]:
# Checking for duplicate rows in each DataFrame
books.duplicated().sum()
users.duplicated().sum()
ratings.duplicated().sum()

0

## Popularity based recommender system

In [12]:
# Creating a popularity-based recommender system
# Merging ratings with books on the ISBN column
ratings_with_name = ratings.merge(books, on='ISBN')
ratings_with_name.shape
ratings_with_name.head()
# If you look at the shape of the new dataframe, rows reduced because there are some books which are in ratings but not in books.

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [13]:
# Finding the count of ratings for each book
num_rating_df = ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating': 'num_rating'}, inplace=True)
num_rating_df

Book-Title  num_rating
0        A Light in the Storm: The Civil War Diary of ...           4
1                                   Always Have Popsicles           1
2                    Apple Magic (The Collector's series)           1
3        Ask Lily (Young Women of Faith: Lily Series, ...           1
4        Beyond IBM: Leadership Marketing and Finance ...           1
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.           2
241067                     Ã?Â?rger mit Produkt X. Roman.           4
241068                                Ã?Â?sterlich leben.           1
241069                              Ã?Â?stlich der Berge.           3
241070                                  Ã?Â?thique en toc           2

[241071 rows x 2 columns]

In [14]:
# Checking the data type of the Book-Rating column and calculating the average rating
ratings_with_name['Book-Rating'].dtype
ratings_with_name['Book-Rating'].mean()

2.839051298761754

In [15]:
# Finding the average rating for each book
avg_rating_df = ratings_with_name.groupby('Book-Title')['Book-Rating'].mean().reset_index()
avg_rating_df.rename(columns = {'Book-Rating' : 'avg_rating'},inplace=True)
avg_rating_df


Book-Title  avg_rating
0        A Light in the Storm: The Civil War Diary of ...    2.250000
1                                   Always Have Popsicles    0.000000
2                    Apple Magic (The Collector's series)    0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...    8.000000
4        Beyond IBM: Leadership Marketing and Finance ...    0.000000
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.    0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.    5.250000
241068                                Ã?Â?sterlich leben.    7.000000
241069                              Ã?Â?stlich der Berge.    2.666667
241070                                  Ã?Â?thique en toc    4.000000

[241071 rows x 2 columns]

In [16]:
# Merging the number of ratings and average ratings DataFrames
popular_df = num_rating_df.merge(avg_rating_df, on='Book-Title')
popular_df


Book-Title  num_rating  \
0        A Light in the Storm: The Civil War Diary of ...           4   
1                                   Always Have Popsicles           1   
2                    Apple Magic (The Collector's series)           1   
3        Ask Lily (Young Women of Faith: Lily Series, ...           1   
4        Beyond IBM: Leadership Marketing and Finance ...           1   
...                                                   ...         ...   
241066                                      Ã?Â?lpiraten.           2   
241067                     Ã?Â?rger mit Produkt X. Roman.           4   
241068                                Ã?Â?sterlich leben.           1   
241069                              Ã?Â?stlich der Berge.           3   
241070                                  Ã?Â?thique en toc           2   

        avg_rating  
0         2.250000  
1         0.000000  
2         0.000000  
3         8.000000  
4         0.000000  
...            ...  
241066    0.000000  
241067    5.250000  
241068    7.000000  
241069    2.666667  
241070    4.000000  

[241071 rows x 3 columns]

In [17]:
# Filtering books with at least 250 ratings and sorting them by average rating
popular_df = popular_df[popular_df['num_rating']>=250].sort_values('avg_rating', ascending=False).head(50)


In [18]:
# Merging with the books DataFrame to get additional book details and removing duplicates
popular_df = popular_df.merge(books, on = 'Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','num_rating','avg_rating']]
popular_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  \
0             J. K. Rowling   
3             J. K. Rowling   
5             J. K. Rowling   
9             J. K. Rowling   
13            J. K. Rowling   
16           J.R.R. TOLKIEN   
17           J.R.R. TOLKIEN   
26            J. K. Rowling   
28           J.R.R. TOLKIEN   
39               Harper Lee   
47                Dan Brown   
53              Mitch Albom   
55            J.D. Salinger   
62             Alice Sebold   
63            George Orwell   
72       Barbara Kingsolver   
73              Neil Gaiman   
78            Sue Monk Kidd   
84            Michael Moore   
85              MITCH ALBOM   
88            Anita Diamant   
89            Douglas Adams   
98       Barbara Ehrenreich   
100            Jon Krakauer   
103            Ray Bradbury   
116         Tracy Chevalier   
117          DIANA GABALDON   
123            Billie Letts   
124       LAURA HILLENBRAND   
126             Yann Martel   
130      Barbara Kingsolver   
133             Dave Pelzer   
135           Frank McCourt   
137         Jennifer Weiner   


## Collaborative Filtering Based Recommender System

In [19]:
## Collaborative Filtering Based Recommender System

# Filtering users who have rated more than 200 books
x = ratings_with_name.groupby('User-ID').count()['Book-Rating']>200
padhe_likhe_users = x[x].index


In [20]:
# Filtering ratings of the users who have rated more than 200 books
filtered_rating = ratings_with_name[ratings_with_name['User-ID'].isin(padhe_likhe_users)]

In [21]:
# Filtering books with at least 50 ratings
y = filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books = y[y].index


In [22]:
# Final filtered ratings DataFrame with only famous books
final_ratings = filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]


In [23]:
# Creating a pivot table with Book-Title as rows and User-ID as columns
pt = final_ratings.pivot_table(index= 'Book-Title', columns='User-ID',values='Book-Rating')
pt


User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [24]:
# Filling NaN values with 0 in the pivot table
pt.fillna(0, inplace=True)

In [26]:
# Importing cosine_similarity from sklearn
from sklearn.metrics.pairwise import cosine_similarity


In [27]:
# Calculating the cosine similarity between books
cosine_similarity(pt).shape


(706, 706)

In [28]:
# Calculating the similarity scores
similarity_scores = cosine_similarity(pt)
similarity_scores


array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [29]:
# Defining the recommend function to get book recommendations based on similarity
def recommend(book_name):
    #index fetch
    index = np.where(pt.index==book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_scores[index])),key=lambda x:x[1],reverse=True)[1:6]
    
    for i in similar_items:
        print(pt.index[i[0]])


In [30]:
# Getting recommendations for the book 'A Bend in the Road'
recommend('A Bend in the Road')


A Walk to Remember
The Last Time They Met : A Novel
Sea Glass: A Novel
Angels
Family Album


In [31]:
recommend('Harry Potter and the Prisoner of Azkaban (Book 3)')

Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Chamber of Secrets (Book 2)
Harry Potter and the Order of the Phoenix (Book 5)
Harry Potter and the Sorcerer's Stone (Book 1)
Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))


In [32]:
recommend('1984')

Animal Farm
The Handmaid's Tale
Brave New World
The Vampire Lestat (Vampire Chronicles, Book II)
The Hours : A Novel


In [33]:
recommend('Into the Wild')

A Widow for One Year
A Man in Full
A Year in Provence
Moo
Good Omens


### Calculating RMSE for the Recommendation Model using Cosine Similarity

In [34]:
from sklearn.model_selection import train_test_split

# Split the data into training and test sets
train_data, test_data = train_test_split(final_ratings, test_size=0.2, random_state=42)


In [35]:
# Create a pivot table for the training set
train_pivot = train_data.pivot_table(index='Book-Title', columns='User-ID', values='Book-Rating')
train_pivot.fillna(0, inplace=True)

In [36]:
# Calculate similarity scores using cosine similarity
similarity_scores_cosine = cosine_similarity(train_pivot)

In [37]:
def predict_ratings(user_id, book_title, similarity_scores, train_pivot):
    if book_title not in train_pivot.index:
        return 0  
    if user_id not in train_pivot.columns:
        return 0 
    
    # Find the index of the book
    book_index = train_pivot.index.get_loc(book_title)
    
    # Calculate the weighted sum of ratings
    similarity_scores_book = similarity_scores[book_index]
    user_ratings = train_pivot.loc[:, user_id]
    
    # Handle case where user has no ratings
    if user_ratings.sum() == 0:
        return user_ratings.mean()
    
    weighted_sum = sum(similarity_scores_book * user_ratings)
    sim_sum = sum(similarity_scores_book)
    
    # Handle division by zero
    if sim_sum == 0:
        return user_ratings.mean()
    
    predicted_rating = weighted_sum / sim_sum
    return predicted_rating


In [38]:
from sklearn.metrics import mean_squared_error

# Predict ratings for the test set using cosine similarity
test_data['predicted_rating'] = test_data.apply(
    lambda x: predict_ratings(x['User-ID'], x['Book-Title'], similarity_scores_cosine, train_pivot), 
    axis=1
)

# Calculate RMSE for cosine similarity
rmse_cosine = np.sqrt(mean_squared_error(test_data['Book-Rating'], test_data['predicted_rating']))
print(f'RMSE using Cosine Similarity: {rmse_cosine}')


RMSE using Cosine Similarity: 3.8156952772541426


### Calculating RMSE for the Recommendation Model using Euclidean Distance

In [39]:
from sklearn.metrics.pairwise import pairwise_distances

# Calculate similarity scores using Euclidean distance
similarity_scores_euclidean = 1 / (1 + pairwise_distances(train_pivot, metric='euclidean'))


In [40]:
# Predict ratings for the test set using Euclidean distance
test_data['predicted_rating_euclidean'] = test_data.apply(lambda x: predict_ratings(x['User-ID'], x['Book-Title'], similarity_scores_euclidean, train_pivot), axis=1)


In [41]:
# Calculate RMSE
rmse_euclidean = np.sqrt(mean_squared_error(test_data['Book-Rating'], test_data['predicted_rating_euclidean']))
print(f'RMSE using Euclidean Distance: {rmse_euclidean}')


RMSE using Euclidean Distance: 3.867942020595821


### Results Comparison

- **RMSE using Cosine Similarity**: 3.898
- **RMSE using Euclidean Distance**: 3.952

### Interpretation

The lower RMSE indicates better performance in predicting the ratings. In this case:

- The RMSE for **Cosine Similarity** is slightly lower than for Euclidean Distance, suggesting that the Cosine Similarity metric provides more accurate predictions for this recommendation system.

### Conclusion

Based on the RMSE values, the recommendation model using **Cosine Similarity** performs better compared to using **Euclidean Distance** for predicting book ratings in this dataset.


# Item Based

In [42]:
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity


# Adjust filtering criteria
# Filter users who have rated at least 10 books
user_counts = ratings['User-ID'].value_counts()
active_users = user_counts[user_counts >= 10].index
filtered_ratings = ratings[ratings['User-ID'].isin(active_users)]

# Filter books that have at least 20 ratings
book_counts = filtered_ratings['ISBN'].value_counts()
popular_books = book_counts[book_counts >= 20].index
filtered_ratings = filtered_ratings[filtered_ratings['ISBN'].isin(popular_books)]

# Add the specific book back if it was filtered out
isbn_to_check = '034545104X'
if isbn_to_check not in filtered_ratings['ISBN'].unique():
    specific_book_ratings = ratings[ratings['ISBN'] == isbn_to_check]
    filtered_ratings = pd.concat([filtered_ratings, specific_book_ratings])

# Create user-item matrix
user_item_matrix = filtered_ratings.pivot(index='User-ID', columns='ISBN', values='Book-Rating').fillna(0)
item_user_matrix = user_item_matrix.T

# Calculate cosine similarity
item_similarity = cosine_similarity(item_user_matrix)

# Convert to DataFrame for easy handling
item_similarity_df = pd.DataFrame(item_similarity, index=item_user_matrix.index, columns=item_user_matrix.index)

# Function to get top-N similar books
def get_similar_books(book_isbn, top_n=5):
    if book_isbn not in item_similarity_df.columns:
        print(f"Book ISBN {book_isbn} not found in the dataset.")
        return []
    similar_books = item_similarity_df[book_isbn].sort_values(ascending=False).head(top_n + 1).index.tolist()
    similar_books.remove(book_isbn)  # Remove the book itself
    return similar_books

# Example: Get top 5 books similar to a specific book ISBN
similar_books = get_similar_books(isbn_to_check, top_n=5)
print(f"Books similar to {isbn_to_check}:", similar_books)


Books similar to 034545104X: ['0440236770', '1551668939', '0451192613', '1551667118', '0060001445']


### Without reducing the dimensioanlity 


In [43]:
from sklearn.metrics.pairwise import cosine_similarity
# Transpose the pivot table to have books as columns and users as rows
pt_transposed = pt.T

In [44]:
# Calculate the cosine similarity matrix
similarity_matrix = cosine_similarity(pt_transposed)
similarity_df = pd.DataFrame(similarity_matrix, index=pt.columns, columns=pt.columns)

### Recommendation

In [45]:
def recommend_books(book_title, similarity_df, pt, n=5):
    # Get the index of the book that matches the title
    book_idx = pt.index.get_loc(book_title)
    
    # Get the similarity scores for this book with all other books
    similarity_scores = similarity_df.iloc[book_idx]
    
    # Sort the books by their similarity scores in descending order
    sorted_indices = similarity_scores.argsort()[::-1]
    
    # Get the top n similar books
    top_n_books = pt.index[sorted_indices[1:n+1]]
    return top_n_books

#Get top 5 book recommendations for '1984'
recommendations_without_dim_reduction = recommend_books('1984', similarity_df, pt, n=5)
recommendations_without_dim_reduction

Index(['The Perfect Storm : A True Story of Men Against the Sea',
       'The Clinic (Alex Delaware Novels (Paperback))', 'The Mulberry Tree',
       'No Safe Place', 'The Bourne Identity'],
      dtype='object', name='Book-Title')

## With Dimensionality Reduction

## using svd

In [46]:
from sklearn.decomposition import TruncatedSVD

# Apply SVD to reduce the dimensionality
svd = TruncatedSVD(n_components=20)  # You can choose the number of components
reduced_matrix = svd.fit_transform(pt.T)
reduced_df = pd.DataFrame(reduced_matrix, index=pt.columns)


In [47]:
# Calculate the cosine similarity matrix on the reduced data
reduced_similarity_matrix = cosine_similarity(reduced_df)
reduced_similarity_df = pd.DataFrame(reduced_similarity_matrix, index=pt.columns, columns=pt.columns)


In [48]:
# Reuse the same recommendation function
recommendations_with_dim_reduction = recommend_books('1984', reduced_similarity_df, pt, n=5)
recommendations_with_dim_reduction


Index(['No Safe Place',
       'Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))',
       'Invasion', 'The Perfect Storm : A True Story of Men Against the Sea',
       'The Accidental Tourist'],
      dtype='object', name='Book-Title')

In [51]:
print("Recommendations without Dimensionality Reduction:")
print(recommendations_without_dim_reduction)

print("\nRecommendations with Dimensionality Reduction:")
print(recommendations_with_dim_reduction)


Recommendations without Dimensionality Reduction:
Index(['The Perfect Storm : A True Story of Men Against the Sea',
       'The Clinic (Alex Delaware Novels (Paperback))', 'The Mulberry Tree',
       'No Safe Place', 'The Bourne Identity'],
      dtype='object', name='Book-Title')

Recommendations with Dimensionality Reduction:
Index(['No Safe Place',
       'Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))',
       'Invasion', 'The Perfect Storm : A True Story of Men Against the Sea',
       'The Accidental Tourist'],
      dtype='object', name='Book-Title')


## Compare rmse 

## Ratings Without Dimensionality Reduction:

In [52]:
def predict_ratings(similarity_df, pt):
    mean_user_rating = pt.mean(axis=1).values  # Convert to numpy array
    ratings_diff = (pt.T - mean_user_rating).T
    similarity_sum = np.array([np.abs(similarity_df).sum(axis=1)]).T
    pred = mean_user_rating[:, np.newaxis] + (similarity_df.dot(ratings_diff.T).T / similarity_sum.T)
    return pred

pred_without_dim_reduction = predict_ratings(similarity_df, pt)

## Ratings With Dimensionality Reduction:

In [53]:
pred_with_dim_reduction = predict_ratings(reduced_similarity_df, pt)

In [54]:
def calculate_rmse(predictions, actuals):
    # Convert DataFrames to NumPy arrays
    predictions = predictions.to_numpy()
    actuals = actuals.to_numpy()
    
    # Get the non-zero indices
    non_zero_indices = actuals.nonzero()
    
    # Extract the predicted and actual values at these indices
    pred_values = predictions[non_zero_indices].flatten()
    actual_values = actuals[non_zero_indices].flatten()
    
    # Calculate the RMSE
    return np.sqrt(((pred_values - actual_values) ** 2).mean())

rmse_without_dim_reduction = calculate_rmse(pd.DataFrame(pred_without_dim_reduction), pt)
rmse_with_dim_reduction = calculate_rmse(pd.DataFrame(pred_with_dim_reduction), pt)

## Compare the result

In [55]:
#RMSE results
print(f"RMSE without Dimensionality Reduction: {rmse_without_dim_reduction}")
print(f"RMSE with Dimensionality Reduction: {rmse_with_dim_reduction}")

RMSE without Dimensionality Reduction: 6.892427666440504
RMSE with Dimensionality Reduction: 7.52996900659038


Item-Based Collaborative Filtering: This method provides more accurate and personalized recommendations, with a lower RMSE of 6.8924, but it is also computationally demanding.
Item-Based Collaborative Filtering with SVD: By reducing data dimensionality, this approach balances accuracy and computational efficiency, with an RMSE of 7.5309, slightly higher than the traditional item-based method
